### PR TITLE
ONNX Export Maxpool Indices

### PR DESCRIPTION
**Pull request by lara-hdr**

### CODE REVIEW

**Review by houseroad — CHANGES_REQUESTED** [review]
The change seems not keeping the semantics of the max_pool_with_indices. Could you explain why you want to flatten the indices? If you export the pytorch model like this, the exported graph won't be correct.

**Comment by lara-hdr** [comment]
@houseroad , in ONNX, for the Maxpool operator the indices are computed as flatten 1-D tensor (https://github.com/onnx/onnx/blob/master/docs/Operators.md#MaxPool). 
Without this conversion, the output values won't be matching in both models.

**Comment by houseroad** [comment]
Please check the doc here: https://github.com/pytorch/pytorch/issues/16742
```
Indices tensor from max pooling across the input tensor. The dimensions of indices are the same as output tensor. The values in indices of are the indices of the selected values during pooling. The indices are computed as flatten 1-D tensor, and the indices do not consider padding. So the values in indices are in [0, N x C x D1 x ... x Dn).
```

The dimension is the same as the dimension of the first output. Only the indices are computed as flatten 1d tensor. Actually, ONNX is aligned with PyTorch.

Also I am wondering whether we can discard the second output if it's not used. (Right now, it starts causing the caffe2 failing.)

**Comment by lara-hdr** [comment]
This is a simple example to better explain the issue.

For **Maxpool1d(kernel=1,stride=1,return_indices=True)**, with the input **torch.ones(1,2,2)**. The computed indices are the following.
- output **indices pytorch** :
[[0,1],
 [0,1]]
 - output **indices onnx**:
[[0,1],
 [2,3]]

In the added code, my purpose was to convert the indices from one format to the other to be able to match the results.

So **flattened_indices** will have the value of each index and will be equal to :
[[0,1],
 [2,3]]

Then **s** will get the first value of each line (so 0 and 2).

And the subtraction executes : 
[[0-0,1-0],
 [2-2,3-2]]

So **indices** results to the expected output which is  :
[[0,1],
 [0,1]]


Definitely discarding the second output if it's not used would be great. However, I am not sure were to make this change since the functions in symbolic.py do not have the "return_indices" param. Could you point me to the right file to modify to do so?
Thanks

**Comment by ezyang** [comment]
ping @houseroad

**Comment by houseroad** [comment]
We had discussion offline. We will update the MaxPool and MaxUnpool in ONNX. But for current version, we need more change to make sure the test case passing.

**Comment by houseroad** [comment]
For now, let's put this PR on hold.

File: torch/onnx/symbolic.py [hunk 1]
```diff
@@ -595,11 +603,19 @@ def max_pool2d_with_indices(g, input, kernel_size, stride, padding, dilation, ce
...
+    r, indices = g.op("MaxPool", input, outputs=2,
+                      kernel_shape_i=_pair(kernel_size),
+                      pads_i=_pair(padding) * 2,
+                      strides_i=_pair(stride))
+    # easy but hacky way to get flattened indices values
+    # to be used to convert the indices values to non-flattened
+    _, flattened_indices = g.op("MaxPool", input, outputs=2,
+                                kernel_shape_i=[1, 1],
+                                strides_i=[1, 1])
+    # convert indices to have non-flattened indices values
```
Comment (houseroad) [inline torch/onnx/symbolic.py]: Could you add more comments on how to compute the non-flattened indices? It's tricky, and not easy to understand.

**Review by houseroad — COMMENTED** [review]
Looks good, but please add more comments about how to compute non-flatten indices.

File: torch/onnx/symbolic.py [hunk 2]
```diff
@@ -580,11 +580,30 @@ def max_pool1d_with_indices(g, input, kernel_size, stride, padding, dilation, ce
...
-    r = g.op("MaxPool", input,
-             kernel_shape_i=_single(kernel_size),
-             pads_i=_single(padding) * 2,
-             strides_i=_single(stride))
-    return r, None
+    r, indices = g.op("MaxPool", input, outputs=2,
+                      kernel_shape_i=_single(kernel_size),
+                      pads_i=_single(padding) * 2,
+                      strides_i=_single(stride))
+    # easy but hacky way to get flattened indices values
```
Comment (houseroad) [inline torch/onnx/symbolic.py]: Let's just keep the comment at one place, and refer it at other places.

**Review by houseroad — COMMENTED** [review]
Since the PR is created 20+ days ago, let's rebase to master. Otherwise, CI may give many false positive signals.

**Review by houseroad — APPROVED** [review]
Thanks a lot for the contribution. Looks good now!

**Comment by ananiask8** [comment]
> This is a simple example to better explain the issue.
> 
> For **Maxpool1d(kernel=1,stride=1,return_indices=True)**, with the input **torch.ones(1,2,2)**. The computed indices are the following.
> 
> * output **indices pytorch** :
>   [[0,1],
>   [0,1]]
> * output **indices onnx**:
>   [[0,1],
>   [2,3]]
> 
> In the added code, my purpose was to convert the indices from one format to the other to be able to match the results.
> 
> So **flattened_indices** will have the value of each index and will be equal to :
> [[0,1],
> [2,3]]
> 
> Then **s** will get the first value of each line (so 0 and 2).
> 
> And the subtraction executes :
> [[0-0,1-0],
> [2-2,3-2]]
> 
> So **indices** results to the expected output which is :
> [[0,1],
> [0,1]]
> 
> Definitely.. Discarding the second output if it's not used would be great. However, I am not sure were to make this change since the functions in symbolic.py do not have the "return_indices" param. Could you point me to the right file to modify to do so?
> Thanks

Hello @lara-hdr . Is there any way that I could use the indices returned and have an output with the same size as the input, with the frequencies that each index appears?

I think I could do it with the original indices that ONNX returns, and using `bincount`. Is there a way to get these indices in a way that is efficient in `cuda`?

Also, if there is any alternative way to do it?

For a more specific explanation, see [here](https://discuss.pytorch.org/t/max-min-pooling-winner-counter-layer/41500).

**Pull request closed without merging**